File: youtube_downloader.js
#!/usr/bin/env bun

/**
 * Script para baixar vídeos do YouTube com faixas de áudio específicas
 * Requer: bun e yt-dlp instalados
 */

import { spawn } from 'child_process';
import { mkdir } from 'fs/promises';
import { existsSync } from 'fs';
import { join } from 'path';
import readline from 'readline';

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

// Função para executar comandos e retornar a saída
async function executarComando(comando, args) {
  return new Promise((resolve, reject) => {
    const processo = spawn(comando, args);
    let saida = '';
    let erro = '';

    processo.stdout.on('data', (dados) => {
      saida += dados.toString();
    });

    processo.stderr.on('data', (dados) => {
      erro += dados.toString();
    });

    processo.on('close', (codigo) => {
      if (codigo !== 0) {
        reject(new Error(`Comando falhou com código ${codigo}: ${erro}`));
      } else {
        resolve(saida);
      }
    });
  });
}

// Função para perguntar ao usuário
function perguntar(pergunta) {
  return new Promise((resolve) => {
    rl.question(pergunta, (resposta) => {
      resolve(resposta);
    });
  });
}

// Função para obter informações do vídeo
async function obterInfoVideo(url) {
  try {
    const saida = await executarComando('yt-dlp', ['-j', url]);
    return JSON.parse(saida);
  } catch (erro) {
    console.error('Erro ao obter informações do vídeo:', erro.message);
    process.exit(1);
  }
}

// Função para mostrar formatos de áudio disponíveis
function mostrarFormatosAudio(formatos) {
  console.log('\nFaixas de áudio disponíveis:');
  console.log('ID\tExtensão\tCodec\t\tBitrate\tIdioma/Descrição');
  console.log('-'.repeat(80));

  const formatosAudio = formatos.filter(formato => 
    formato.vcodec === 'none' || 
    (formato.acodec !== 'none' && formato.format && formato.format.includes('audio only'))
  );

  formatosAudio.forEach(formato => {
    const idioma = formato.language || 'desconhecido';
    const descricao = formato.format_note || '';
    console.log(`${formato.format_id}\t${formato.ext || 'N/A'}\t\t${formato.acodec || 'N/A'}\t\t${formato.abr || 'N/A'}k\t${idioma} ${descricao}`);
  });

  return formatosAudio;
}

// Função para mostrar formatos de vídeo disponíveis
function mostrarFormatosVideo(formatos) {
  console.log('\nFormatos de vídeo disponíveis (sem áudio):');
  console.log('ID\tExtensão\tResolução\tCodec\t\tBitrate');
  console.log('-'.repeat(80));

  const formatosVideo = formatos.filter(formato => 
    formato.acodec === 'none' && formato.vcodec !== 'none'
  );

  formatosVideo.forEach(formato => {
    console.log(`${formato.format_id}\t${formato.ext || 'N/A'}\t\t${formato.resolution || 'N/A'}\t\t${formato.vcodec || 'N/A'}\t\t${formato.vbr || 'N/A'}k`);
  });

  return formatosVideo;
}

// Função para encontrar a melhor faixa de áudio no idioma preferido
function encontrarMelhorAudio(formatosAudio, idiomaPreferido = 'pt-BR') {
  // Filtrar por idioma preferido
  let audiosIdioma = formatosAudio.filter(formato => 
    formato.language && formato.language.toLowerCase().includes(idiomaPreferido.toLowerCase())
  );
  
  // Se não encontrar no idioma preferido, use qualquer um
  if (audiosIdioma.length === 0) {
    audiosIdioma = formatosAudio;
  }
  
  // Procurar por áudios de qualidade média ou alta
  const audiosMedium = audiosIdioma.filter(formato => 
    formato.format_note && formato.format_note.toLowerCase().includes('medium')
  );
  
  if (audiosMedium.length > 0) {
    // Preferir m4a sobre webm para compatibilidade
    const m4aAudio = audiosMedium.find(audio => audio.ext === 'm4a');
    if (m4aAudio) {
      return m4aAudio.format_id;
    }
    // Se não encontrar m4a, use o primeiro de qualidade média
    return audiosMedium[0].format_id;
  }
  
  // Se não encontrar de qualidade média, use o de maior bitrate
  audiosIdioma.sort((a, b) => (b.abr || 0) - (a.abr || 0));
  if (audiosIdioma.length > 0) {
    return audiosIdioma[0].format_id;
  }
  
  // Se tudo falhar, retorne null
  return null;
}

// Função para encontrar o melhor formato de vídeo
function encontrarMelhorVideo(formatosVideo, resolucaoMaxima = 1080) {
  // Filtrar por resolução máxima
  const videosFiltrados = [];
  
  for (const formato of formatosVideo) {
    const resolucao = formato.resolution;
    if (typeof resolucao === 'string' && resolucao.includes('x')) {
      const altura = parseInt(resolucao.split('x')[1]);
      if (altura <= resolucaoMaxima) {
        videosFiltrados.push([formato, altura]);
      }
    }
  }
  
  // Ordenar por altura (resolução) em ordem decrescente
  videosFiltrados.sort((a, b) => b[1] - a[1]);
  
  if (videosFiltrados.length > 0) {
    // Pegar o vídeo com maior resolução
    const melhorVideo = videosFiltrados[0][0];
    
    // Preferir codecs mais eficientes (AV1 > VP9 > H.264)
    for (const [video, _] of videosFiltrados) {
      if (video.height === melhorVideo.height) {
        const codec = (video.vcodec || '').toLowerCase();
        if (codec.includes('av01')) {  // AV1
          return video.format_id;
        }
      }
    }
    
    // Se não encontrar AV1, use o de maior resolução
    return melhorVideo.format_id;
  }
  
  // Se tudo falhar, retorne null
  return null;
}

// Função para baixar o vídeo
async function baixarVideo(url, idVideo, idAudio, pastaDestino) {
  console.log(`\nBaixando vídeo com formato de vídeo ${idVideo} e áudio ${idAudio}...`);
  
  try {
    await executarComando('yt-dlp', [
      '-f', `${idVideo}+${idAudio}`,
      '-o', `${pastaDestino}/%(title)s.%(ext)s`,
      url
    ]);
    console.log(`\nDownload concluído! O arquivo foi salvo na pasta '${pastaDestino}'.`);
  } catch (erro) {
    console.error('Erro ao baixar o vídeo:', erro.message);
    process.exit(1);
  }
}

// Função principal
async function main() {
  if (process.argv.length < 3) {
    console.log('Uso: bun youtube_downloader.js URL_DO_VIDEO');
    process.exit(1);
  }

  const url = process.argv[2];
  const info = await obterInfoVideo(url);
  
  console.log(`Título do vídeo: ${info.title}`);
  
  const formatosAudio = mostrarFormatosAudio(info.formats);
  const formatosVideo = mostrarFormatosVideo(info.formats);
  
  if (formatosAudio.length === 0) {
    console.log('Nenhuma faixa de áudio encontrada.');
    process.exit(1);
  }
  
  if (formatosVideo.length === 0) {
    console.log('Nenhum formato de vídeo encontrado.');
    process.exit(1);
  }
  
  // Opção automática ou manual
  const escolha = await perguntar('\nDeseja escolher automaticamente a melhor qualidade? (S/n): ');
  
  let idAudio, idVideo;
  
  if (escolha.trim().toLowerCase() === '' || escolha.trim().toLowerCase() === 's') {
    // Escolha automática
    const idiomaPreferido = await perguntar('Digite o idioma preferido (deixe em branco para português do Brasil): ');
    const idioma = idiomaPreferido.trim() || 'pt-BR';
    
    const resolucaoMaximaInput = await perguntar('Digite a resolução máxima desejada (deixe em branco para 1080p): ');
    let resolucaoMaxima = 1080;
    
    if (resolucaoMaximaInput.trim()) {
      const resolucaoInt = parseInt(resolucaoMaximaInput);
      if (!isNaN(resolucaoInt)) {
        resolucaoMaxima = resolucaoInt;
      } else {
        console.log('Resolução inválida, usando 1080p como padrão.');
      }
    }
    
    idAudio = encontrarMelhorAudio(formatosAudio, idioma);
    idVideo = encontrarMelhorVideo(formatosVideo, resolucaoMaxima);
    
    if (!idAudio) {
      console.log('Não foi possível encontrar uma faixa de áudio adequada.');
      process.exit(1);
    }
    
    if (!idVideo) {
      console.log('Não foi possível encontrar um formato de vídeo adequado.');
      process.exit(1);
    }
    
    console.log('\nSelecionado automaticamente:');
    console.log(`- Áudio: ${idAudio}`);
    console.log(`- Vídeo: ${idVideo}`);
  } else {
    // Escolha manual
    idAudio = await perguntar('\nDigite o ID da faixa de áudio desejada: ');
    idVideo = await perguntar('Digite o ID do formato de vídeo desejado: ');
    
    // Verificar se os IDs existem
    const audioValido = formatosAudio.some(f => f.format_id === idAudio);
    const videoValido = formatosVideo.some(f => f.format_id === idVideo);
    
    if (!audioValido) {
      console.log(`ID de áudio '${idAudio}' não encontrado.`);
      process.exit(1);
    }
    
    if (!videoValido) {
      console.log(`ID de vídeo '${idVideo}' não encontrado.`);
      process.exit(1);
    }
  }
  
  // Criar pasta de destino se não existir
  const pastaDestino = 'downloads';
  if (!existsSync(pastaDestino)) {
    await mkdir(pastaDestino);
  }
  
  // Baixar o vídeo
  await baixarVideo(url, idVideo, idAudio, pastaDestino);
  
  rl.close();
}

main().catch(erro => {
  console.error('Erro:', erro);
  process.exit(1);
}); 

File: youtube_downloader.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import subprocess
import sys
import json
import os
import re

def obter_formatos(url):
    """Obtém todos os formatos disponíveis para o vídeo."""
    comando = ["yt-dlp", "-j", url]
    resultado = subprocess.run(comando, capture_output=True, text=True)
    
    if resultado.returncode != 0:
        print(f"Erro ao obter informações do vídeo: {resultado.stderr}")
        sys.exit(1)
    
    info = json.loads(resultado.stdout)
    return info.get("formats", []), info.get("title", "Video")

def mostrar_formatos_audio(formatos):
    """Mostra apenas os formatos de áudio disponíveis."""
    print("\nFaixas de áudio disponíveis:")
    print("ID\tExtensão\tCodec\t\tBitrate\tIdioma/Descrição")
    print("-" * 80)
    
    formatos_audio = []
    for formato in formatos:
        if formato.get("vcodec") == "none" or formato.get("acodec") != "none" and "audio only" in formato.get("format", ""):
            idioma = formato.get("language", "")
            if not idioma:
                idioma = "desconhecido"
            
            descricao = formato.get("format_note", "")
            
            print(f"{formato['format_id']}\t{formato.get('ext', 'N/A')}\t\t{formato.get('acodec', 'N/A')}\t\t{formato.get('abr', 'N/A')}k\t{idioma} {descricao}")
            formatos_audio.append(formato)
    
    return formatos_audio

def mostrar_formatos_video(formatos):
    """Mostra apenas os formatos de vídeo (sem áudio) disponíveis."""
    print("\nFormatos de vídeo disponíveis (sem áudio):")
    print("ID\tExtensão\tResolução\tCodec\t\tBitrate")
    print("-" * 80)
    
    formatos_video = []
    for formato in formatos:
        if formato.get("acodec") == "none" and formato.get("vcodec") != "none":
            print(f"{formato['format_id']}\t{formato.get('ext', 'N/A')}\t\t{formato.get('resolution', 'N/A')}\t\t{formato.get('vcodec', 'N/A')}\t\t{formato.get('vbr', 'N/A')}k")
            formatos_video.append(formato)
    
    return formatos_video

def encontrar_melhor_audio(formatos_audio, idioma_preferido="pt-BR"):
    """Encontra a melhor faixa de áudio no idioma preferido."""
    # Filtrar por idioma preferido
    audios_idioma = [f for f in formatos_audio if idioma_preferido.lower() in str(f.get("language", "")).lower()]
    
    # Se não encontrar no idioma preferido, use qualquer um
    if not audios_idioma:
        audios_idioma = formatos_audio
    
    # Procurar por áudios de qualidade média ou alta
    audios_medium = [f for f in audios_idioma if "medium" in str(f.get("format_note", "")).lower()]
    if audios_medium:
        # Preferir m4a sobre webm para compatibilidade
        for audio in audios_medium:
            if audio.get("ext") == "m4a":
                return audio["format_id"]
        # Se não encontrar m4a, use o primeiro de qualidade média
        return audios_medium[0]["format_id"]
    
    # Se não encontrar de qualidade média, use o de maior bitrate
    audios_idioma.sort(key=lambda x: float(x.get("abr", 0) or 0), reverse=True)
    if audios_idioma:
        return audios_idioma[0]["format_id"]
    
    # Se tudo falhar, retorne None
    return None

def encontrar_melhor_video(formatos_video, resolucao_maxima=1080):
    """Encontra o melhor formato de vídeo até a resolução máxima especificada."""
    # Filtrar por resolução máxima
    videos_filtrados = []
    for formato in formatos_video:
        resolucao = formato.get("resolution", "")
        if isinstance(resolucao, str) and "x" in resolucao:
            altura = int(resolucao.split("x")[1])
            if altura <= resolucao_maxima:
                videos_filtrados.append((formato, altura))
    
    # Ordenar por altura (resolução) em ordem decrescente
    videos_filtrados.sort(key=lambda x: x[1], reverse=True)
    
    if videos_filtrados:
        # Pegar o vídeo com maior resolução
        melhor_video = videos_filtrados[0][0]
        
        # Preferir codecs mais eficientes (AV1 > VP9 > H.264)
        for video, _ in videos_filtrados:
            if video["height"] == melhor_video["height"]:
                codec = video.get("vcodec", "").lower()
                if "av01" in codec:  # AV1
                    return video["format_id"]
        
        # Se não encontrar AV1, use o de maior resolução
        return melhor_video["format_id"]
    
    # Se tudo falhar, retorne None
    return None

def baixar_video(url, id_video, id_audio, pasta_destino="."):
    """Baixa o vídeo com a faixa de áudio especificada."""
    comando = [
        "yt-dlp", 
        "-f", f"{id_video}+{id_audio}", 
        "-o", f"{pasta_destino}/%(title)s.%(ext)s",
        url
    ]
    
    print(f"\nBaixando vídeo com formato de vídeo {id_video} e áudio {id_audio}...")
    subprocess.run(comando)

def main():
    if len(sys.argv) < 2:
        print("Uso: python youtube_downloader.py URL_DO_VIDEO")
        sys.exit(1)
    
    url = sys.argv[1]
    formatos, titulo = obter_formatos(url)
    
    print(f"Título do vídeo: {titulo}")
    
    formatos_audio = mostrar_formatos_audio(formatos)
    formatos_video = mostrar_formatos_video(formatos)
    
    if not formatos_audio:
        print("Nenhuma faixa de áudio encontrada.")
        sys.exit(1)
    
    if not formatos_video:
        print("Nenhum formato de vídeo encontrado.")
        sys.exit(1)
    
    # Opção automática ou manual
    escolha = input("\nDeseja escolher automaticamente a melhor qualidade? (S/n): ").strip().lower()
    
    if escolha == "" or escolha == "s":
        # Escolha automática
        idioma_preferido = input("Digite o idioma preferido (deixe em branco para português do Brasil): ").strip()
        if not idioma_preferido:
            idioma_preferido = "pt-BR"
        
        resolucao_maxima = input("Digite a resolução máxima desejada (deixe em branco para 1080p): ").strip()
        if not resolucao_maxima:
            resolucao_maxima = 1080
        else:
            try:
                resolucao_maxima = int(resolucao_maxima)
            except ValueError:
                print("Resolução inválida, usando 1080p como padrão.")
                resolucao_maxima = 1080
        
        id_audio = encontrar_melhor_audio(formatos_audio, idioma_preferido)
        id_video = encontrar_melhor_video(formatos_video, resolucao_maxima)
        
        if not id_audio:
            print("Não foi possível encontrar uma faixa de áudio adequada.")
            sys.exit(1)
        
        if not id_video:
            print("Não foi possível encontrar um formato de vídeo adequado.")
            sys.exit(1)
        
        print(f"\nSelecionado automaticamente:")
        print(f"- Áudio: {id_audio}")
        print(f"- Vídeo: {id_video}")
    else:
        # Escolha manual
        id_audio = input("\nDigite o ID da faixa de áudio desejada: ")
        id_video = input("Digite o ID do formato de vídeo desejado: ")
        
        # Verificar se os IDs existem
        audio_valido = any(f["format_id"] == id_audio for f in formatos_audio)
        video_valido = any(f["format_id"] == id_video for f in formatos_video)
        
        if not audio_valido:
            print(f"ID de áudio '{id_audio}' não encontrado.")
            sys.exit(1)
        
        if not video_valido:
            print(f"ID de vídeo '{id_video}' não encontrado.")
            sys.exit(1)
    
    # Criar pasta de destino se não existir
    pasta_destino = "downloads"
    os.makedirs(pasta_destino, exist_ok=True)
    
    # Baixar o vídeo
    baixar_video(url, id_video, id_audio, pasta_destino)
    print(f"\nDownload concluído! O arquivo foi salvo na pasta '{pasta_destino}'.")

if __name__ == "__main__":
    main() 